src/measure_theory/function/jacobian.lean

```diff
@@ -0,0 +1,1267 @@
+  have E₂ : ∀ᵐ (x : E) ∂μ.restrict (s ∩ t n), g x = A n,
+  { suffices H : ∀ᵐ (x : E) ∂μ.restrict (t n), g x = A n,
+      from ae_mono (restrict_mono (inter_subset_right _ _) le_rfl) H,
+    filter_upwards [ae_restrict_mem (t_meas n)],
+    exact hg n },
+  -- putting these two properties together gives the conclusion.
+  filter_upwards [E₁, E₂],
+  assume x hx1 hx2,
+  rw ← nndist_eq_nnnorm at hx1,
+  rw [hx2, dist_comm],
+  exact hx1,
+end
+
+lemma ae_measurable_of_real_abs_det_fderiv_within
+  {f : E → E} {s : set E} (hs : measurable_set s) {f' : E → (E →L[ℝ] E)}
+  (hf' : ∀ x ∈ s, has_fderiv_within_at f (f' x) s x) :
+  ae_measurable (λ x, ennreal.of_real (|(f' x).det|)) (μ.restrict s) :=
+begin
+  apply ennreal.measurable_of_real.comp_ae_measurable,
+  refine continuous_abs.measurable.comp_ae_measurable _,
+  refine continuous_linear_map.continuous_det.measurable.comp_ae_measurable _,
+  exact ae_measurable_fderiv_within μ hs hf'
+end
+
+lemma ae_measurable_to_nnreal_abs_det_fderiv_within
+  {f : E → E} {s : set E} (hs : measurable_set s) {f' : E → (E →L[ℝ] E)}
+  (hf' : ∀ x ∈ s, has_fderiv_within_at f (f' x) s x) :
+  ae_measurable (λ x, |(f' x).det|.to_nnreal) (μ.restrict s) :=
+begin
+  apply measurable_real_to_nnreal.comp_ae_measurable,
+  refine continuous_abs.measurable.comp_ae_measurable _,
+  refine continuous_linear_map.continuous_det.measurable.comp_ae_measurable _,
+  exact ae_measurable_fderiv_within μ hs hf'
+end
+
+/-- If a function is differentiable and injective on a measurable set,
+then the image is measurable.-/
+lemma measurable_image_of_fderiv_within
+  {f : E → E} {s : set E} (hs : measurable_set s) {f' : E → (E →L[ℝ] E)}
+  (hf' : ∀ x ∈ s, has_fderiv_within_at f (f' x) s x) (hf : inj_on f s) :
+  measurable_set (f '' s) :=
+begin
+  have : differentiable_on ℝ f s := λ x hx, (hf' x hx).differentiable_within_at,
+  exact hs.image_of_continuous_on_inj_on (differentiable_on.continuous_on this) hf,
+end
+
+/-- If a function is differentiable and injective on a measurable set `s`, then its restriction
+to `s` is a measurable embedding. -/
+lemma measurable_embedding_of_fderiv_within
+  {f : E → E} {s : set E} (hs : measurable_set s) {f' : E → (E →L[ℝ] E)}
+  (hf' : ∀ x ∈ s, has_fderiv_within_at f (f' x) s x) (hf : inj_on f s) :
+  measurable_embedding (s.restrict f) :=
+begin
+  have : differentiable_on ℝ f s := λ x hx, (hf' x hx).differentiable_within_at,
+  exact this.continuous_on.measurable_embedding hs hf
+end
+
+/-!
+### Proving the estimate for the measure of the image
+
+We show the formula `∫⁻ x in s, ennreal.of_real (|(f' x).det|) ∂μ = μ (f '' s)`,
+in `lintegral_abs_det_fderiv_eq_add_haar_image`. For this, we show both inequalities in both
+directions, first up to controlled errors and then letting these errors tend to `0`.
+-/
+
+lemma add_haar_image_le_lintegral_abs_det_fderiv_aux1
+  {f : E → E} {s : set E} (hs : measurable_set s)
+  {f' : E → (E →L[ℝ] E)} (hf' : ∀ x ∈ s, has_fderiv_within_at f (f' x) s x)
```

Comment:
`f`, `s` and `f'` appear in every lemma around that point. You could declare them as variables.